modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PageHandler.java

```diff
@@ -0,0 +1,478 @@
+     * @param pageAddr Page address.
+     * @param lsnr     Lock listener.
+     */
+    public static void readUnlock(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            long page,
+            long pageAddr,
+            PageLockListener lsnr
+    ) {
+        lsnr.onReadUnlock(groupId, pageId, page, pageAddr);
+
+        pageMem.readUnlock(groupId, pageId, page);
+    }
+
+    /**
+     * Initializes a new page.
+     *
+     * @param pageMem    Page memory.
+     * @param groupId    Group ID.
+     * @param pageId     Page ID.
+     * @param init       IO for new page initialization.
+     * @param lsnr       Lock listener.
+     * @param statHolder Statistics holder to track IO operations.
+     * @throws IgniteInternalCheckedException If failed.
+     * @see PageIo#initNewPage(long, long, int)
+     */
+    public static void initPage(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            PageIo init,
+            PageLockListener lsnr,
+            IoStatisticsHolder statHolder
+    ) throws IgniteInternalCheckedException {
+        Boolean res = writePage(
+                pageMem,
+                groupId,
+                pageId,
+                lsnr,
+                PageHandler.NO_OP,
+                init,
+                null,
+                0,
+                FALSE,
+                statHolder
+        );
+
+        assert res != FALSE;
+    }
+
+    /**
+     * Executes handler under the read lock or returns {@code lockFailed} if lock failed.
+     *
+     * @param pageMem    Page memory.
+     * @param groupId    Group ID.
+     * @param pageId     Page ID.
+     * @param lsnr       Lock listener.
+     * @param h          Handler.
+     * @param init       IO for new page initialization or {@code null} if it is an existing page.
+     * @param arg        Argument.
+     * @param intArg     Argument of type {@code int}.
+     * @param lockFailed Result in case of lock failure due to page recycling.
+     * @param statHolder Statistics holder to track IO operations.
+     * @return Handler result.
+     * @throws IgniteInternalCheckedException If failed.
+     */
+    public static <X, R> R writePage(
+            PageMemory pageMem,
+            int groupId,
+            final long pageId,
+            PageLockListener lsnr,
+            PageHandler<X, R> h,
+            PageIo init,
+            X arg,
+            int intArg,
+            R lockFailed,
+            IoStatisticsHolder statHolder
+    ) throws IgniteInternalCheckedException {
+        boolean releaseAfterWrite = true;
+        long page = pageMem.acquirePage(groupId, pageId, statHolder);
+        try {
+            long pageAddr = writeLock(pageMem, groupId, pageId, page, lsnr, false);
+
+            if (pageAddr == 0L) {
+                return lockFailed;
+            }
+
+            boolean ok = false;
+
+            try {
+                if (init != null) {
```

Comment:
So we signal that page needs to be initialized by passing `null` in the `init` argument? Or there is some invariant here?

Reply:
Yes, we're passing init IO to initialize page. This is a questionable design choice, I agree, but I won't refactor it right now, we are yet to port code that invokes it